cscs-checks/system/jobreport/gpu_report.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2016-2021 Swiss National Supercomputing Centre (CSCS/ETH Zurich)
+# ReFrame Project Developers. See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+import reframe as rfm
+import reframe.utility.sanity as sn
+import time
+
+from reframe.core.exceptions import SanityError
+from hpctestlib.microbenchmarks.gpu.gpu_burn import GpuBurn
+
+
+@rfm.simple_test
+class gpu_usage_report_check(GpuBurn):
+    '''Check the output from the job report.
+
+    This check uses the gpu burn app and checks that the job report produces
+    senssible gpu usage data. The total number of reported nodes is used as
+    a performance metric, given that the number of reported nodes might be
+    lower than the total number of nodes. The lower threshold of the performace
+    check can be adjusted with the test variable `perf_floor`.
+    '''
+
+    valid_systems = ['daint:gpu', 'dom:gpu']
+    valid_prog_environs = ['PrgEnv-gnu']
+    descr = 'Check GPU usage from job report'
+    gpu_build = 'cuda'
+    modules = ['craype-accel-nvidia60', 'cdt-cuda']
+    num_tasks = 0
+    num_gpus_per_node = 1
+    burn_time = variable(int, value=10)
+    executable_opts = ['-d', f'{burn_time}']
+    perf_floor = variable(float, value=-0.2)
+    tags = {'production'}
+
+    @rfm.run_before('run')
+    def set_launcher_opts(self):
+        '''Make slurm's output unbuffered.
+
+        Without this, the jobreport data gets written into the stdout after the
+        job is completed, causing the sanity function to issue a sanity error.
+        '''
+        self.job.launcher.options = ['-u']
+
+    @rfm.run_before('sanity')
+    def set_sanity_patterns(self):
+        '''Set sanity patterns and wait for the jobreport.
+
+        If a large number of nodes is used, the final jobreport output happens
+        much later after job has already completed (this could be up to 25s).
+        However, this wait might not be necessary for a small node count, so
+        we use the try/except block below to wait only if a first call to the
+        sanity function does not succeed.
+        '''
+
+        super().set_sanity_patterns()
+        self.sanity_patterns = sn.all([
+            self.sanity_patterns, self.gpu_usage_sanity()
+        ])
```

Comment:
Which part of the sanity patterns is expected to fail? That from GPU burn or the one from this test?

Reply:
The one from this test. I've updated the try/except block from below to just test the `self.gpu_usage_sanity()` function instead of the full sanity_patterns.